src/Grpc.AspNetCore.Server/Internal/PipeExtensions.cs

```diff
@@ -68,7 +68,14 @@ public static async Task WriteMessageAsync<TResponse>(this PipeWriter pipeWriter
 
                 if (flush)
                 {
-                    await pipeWriter.FlushAsync();
+                    var flushResult = await pipeWriter.FlushAsync();
+
+                    // Workaround bug where FlushAsync doesn't return IsCanceled = true on request abort.
+                    // https://github.com/dotnet/aspnetcore/issues/40788
+                    if (!flushResult.IsCompleted && serverCallContext.CancellationToken.IsCancellationRequested)
+                    {
+                        throw new OperationCanceledException();
```

Comment:
Does this exception end up surfacing to the user? If so, should we add a message?

Reply:
Yes. Added message.